### ISSUE TITLE
Search regressions in nightly builds of kiwix-serve: (1) punctuation floods the UI (2) Title Search goes to the wrong URL

### ISSUE THREAD

**Issue by holta** [kiwix/kiwix-tools#505]
1. Punctuation floods the search UI, as seen in the screenshot below.

   This is a regression compared to prior versions of kiwix-serve (an example is http://download.kiwix.org/release/kiwix-tools/kiwix-tools_linux-armhf-3.1.2-5.tar.gz which does not have this problem).

2. Title Search links fail.

   Example: when you click on `&lt;b&gt;apple&lt;/b&gt;` at the top of the search dropdown (screenshot below), it does NOT go to http://192.168.0.182/kiwix/wiktionary_en_simple_all_maxi_2021-12/A/apple

   Instead, it erroneously goes to: http://192.168.0.182/kiwix/search?content=wiktionary_en_simple_all_maxi_2021-12&pattern=apple

   Whereas: Full-Text Search link DOES work, leading to http://192.168.0.182/kiwix/search?content=wiktionary_en_simple_all_maxi_2021-12&pattern=apple+ (which correctly shows 67 search result articles, that happen to mention the word "apple").

![image](https://user-images.githubusercontent.com/2458907/145421105-9ad04b79-b455-4209-9193-a2e7a8514f7d.png)

CONTEXT:

- Testing occurred with the very latest http://download.kiwix.org/nightly/2021-12-09/kiwix-tools_linux-armhf-2021-12-09.tar.gz on [IIAB 7.2 Preview 2](https://download.iiab.io) on Raspberry Pi OS Lite Bullseye 2021-10-30, on a Raspberry Pi 4.

- The ZIM file used in the above test was the 46M http://download.kiwix.org/zim/wiktionary/wiktionary_en_simple_all_maxi_2021-12.zim from 2021-12-04.

- [iiab-diagnostics](https://github.com/iiab/iiab/blob/master/scripts/iiab-diagnostics.README.md) output: http://sprunge.us/b9m0GQ?en

**Comment by kelson42** [comment]
@holta Thank you very much for the bug report. We should clarify if this is a problem with the ZIM or with the software.

**Comment by holta** [comment]
FYI I just clarified and revised the [above explanation](https://github.com/kiwix/kiwix-tools/issues/505#issue-1075682864), providing a bit more context.

**Comment by kelson42** [comment]
@holta I can not reproduce the problem with the very same ZIM file, page and version of nightly (but built for x86-64).

![image](https://user-images.githubusercontent.com/1029718/145678429-ad749bea-e9e6-42a1-a073-e99f591634d4.png)

Tested with both Chrome & FF.

Please secure:
* You cache is empty
* You test without the reverse proxy of IIAB, so directly with kiwix-serve.

**Comment by holta** [comment]
Clearing the browser cache worked to solve both problems!

- Does anybody have an idea how difficult it would be to bypass the browser cache automatically?

- So that kiwix-serve / kiwix-search stuff is never cached?

- And where would be the best place to do that?

(To keep regular users from getting very confused.)

**Comment by holta** [comment]
CONTEXT: asking regular users to clear their browser cache every time they search Kiwix &mdash; will certainly not work in low-literacy societies especially :smile:

(So if there's anything Kiwix and/or IIAB should do to solve this at the root, together we definitely should!)

**Comment by holta** [comment]
I forgot to mention a 3rd UX bug, that is also quite common, and likewise very frustrating:

_Oftentimes kiwix-serve's search dropdown never appears when you type into the textfield._

_No matter how long you wait after typing in a keyword (e.g. "apple")._

(In any case, this frustrating end-user dilemma also goes away when the browser cache is cleared.  I'll keep an eye on it, to try to understand any underlying patterns.)

**Comment by kelson42** [comment]
@holta One topic per ticket. Please open a new ticket if you think there is an other bug.

**Comment by kelson42** [comment]
> CONTEXT: asking regular users to clear their browser cache every time they search Kiwix — will certainly not work in low-literacy societies especially smile

Agree, there is a weakness in kiwix-serve caching strategy since always. We will fix it, see #650.

**Comment by tim-moody** [comment]
Even if the issue is browser cache, I still wonder where the `<b>apple</b>` got converted to `&lt;b&gt;apple`

Doesn't seem like something the cache would do. But browsers keep trying to be more 'secure', so perhaps we need to know which version of the browsers are used by @kelson42 and @holta

**Comment by holta** [comment]
> Doesn't seem like something the cache would do.

Confusing indeed.  (Certainly clearing the cache made the issue go away, in my testing anyway.)

PS All above tests were done using Chrome 96 as the browser.

**Comment by veloman-yunkan** [comment]
> Even if the issue is browser cache, I still wonder where the `<b>apple</b>` got converted to `&lt;b&gt;apple`

@tim-moody 
Obviously, the way that HTML content is handled by the front-end and back-end of the kiwix-serve suggestion machinery was changed. In this case the new front-end used cached results that have been served by the old backend or something like that.

**Comment by tim-moody** [comment]
on a nuc with recent but not latest kiwix-serve I search for Clouston's hidrotic ectodermal dysplasia on port 3000 (no proxy) and I get the same thing in the drop down, but if I then press down arrow the text in the search box changes to Clouston&apos;s hidrotic ectodermal dysplasia. With up arrow it reverts to the original. I have never done this search before and get the same result on FF, Edge, and Chrome
Kiwix is dated June 9, 2021

I speculate that the search box is being populated with the formatted title from the search dropdown with the formatting changed to url safe.

**Comment by mgautierfr** [comment]
As @veloman-yunkan suggest. It is a problem of compatibility between the frontend and the backend.

> Clearing the browser cache worked to solve both problems!
    Does anybody have an idea how difficult it would be to bypass the browser cache automatically?
    So that kiwix-serve / kiwix-search stuff is never cached?
    And where would be the best place to do that?

What we are caching is not the search/suggestion/random results but the front js. If we change the way we return results to the front (and the way it is handled by the front), user need to get the new front version.
The best way to do it is probably to version the front js/html/css and add this version in the urls. Then it would be seen as different resources by the browser and it would not use a old version of the front.

**Comment by holta** [comment]
1) Thank you @veloman-yunkan and @mgautierfr for confirming.  That helps a lot!

2) In addition, there are also some "excess punctuation + HTML tags" bugs, even when browser cache is cleared.

   @tim-moody tried to give an example above (yesterday) but his posting was not clear.  What he meant to say was:

   > on a nuc with recent but not latest kiwix-serve I search for Clouston's hidrotic ectodermal dysplasia on port 3000 (no proxy) and I get the same thing in the drop down, but if I then press down arrow the text in the search box changes to `Clouston&apos;s hidrotic ectodermal dysplasia`. With up arrow it reverts to the original. I have never done this search before and get the same result on FF, Edge, and Chrome

   > Kiwix is dated June 9, 2021

   (That's kiwix-tools 3.1.2-5 from http://download.kiwix.org/release/kiwix-tools/kiwix-tools_linux-x86_64-3.1.2-5.tar.gz)

   > I speculate that the search box is being populated with the formatted title from the search dropdown with the formatting changed to url safe.

**Comment by kelson42** [comment]
> 2\. In addition, there are also some "excess punctuation + HTML tags" bugs, even when browser cache is cleared.

Please open one new ticket per problem, without bug report, they might never be fixed.

**Comment by holta** [comment]
> one new ticket per problem

Indeed this is the original problem described in the subject line of this ticket (excess punctuation, that is accidentally polluting Kiwix search UX).

**Comment by tim-moody** [comment]
OK. I didn't know what you meant by 'front end' This I understand

>What we are caching is not the search/suggestion/random results but the front js. If we change the way we return results to the front (and the way it is handled by the front), user need to get the new front version.
The best way to do it is probably to version the front js/html/css and add this version in the urls. Then it would be seen as different resources by the browser and it would not use a old version of the front.

Still odd that I experienced it in private mode on FF and even after I cleared the cache.
